### ISSUE TITLE
Allow building OrderList within Strategy

### ISSUE THREAD

**Issue by rsmb7z** [nautechsystems/nautilus_trader#998]
# Feature Request

At the moment you can create Primary (single) order using order factory and the Bracket order as `OrderList`. However building `OrderList` based on two or more primary orders is not possible. It would be more flexible to allow building `OrderList`. I think exposing `self._order_list_id_generator.generate()` and `self._order_id_generator.generate()` to python would be sufficient to achieve this.

**Comment by cjdsellers** [comment]
Hi @rsmb7z 

It is actually possible to add as many orders as you like to an `OrderList` and send them onto `RiskEngine` -> `ExecEngine` -> `ExecClient`. 

How they are then handled is implementation specific at the adapter, and depends if the venue/broker accepts contingent linked orders, and/or order bulks.

An example is the Binance adapter, which doesn't currently use the Binance OCO API, and merely sends one order after the other in a synchronous way:
```python
    async def _submit_order_list(self, command: SubmitOrderList) -> None:
        for order in command.order_list:
            self.generate_order_submitted(
                strategy_id=order.strategy_id,
                instrument_id=order.instrument_id,
                client_order_id=order.client_order_id,
                ts_event=self._clock.timestamp_ns(),
            )

        for order in command.order_list:
            if order.linked_order_ids:  # TODO(cs): Implement
                self._log.warning(f"Cannot yet handle OCO conditional orders, {order}.")
            await self._submit_order(order)
```

We'll eventually integrate [Binance OCOs.](https://binance-docs.github.io/apidocs/spot/en/#new-oco-trade)

When you mention primary orders, are you referring to say parent orders in an OTO arrangement?

Also, could you be more specific on how exposing the ID generators could help here?

**Comment by rsmb7z** [comment]
Hi @cjdsellers 
The above example is from the adapter side, and yes `OrderList`, as core functionality is there using bracket order, as in this example strategy:
https://github.com/nautechsystems/nautilus_trader/blob/master/nautilus_trader/examples/strategies/ema_cross_bracket.py#L226

In `self.order_factory.bracket` case however, you always have to specifiy 3 orders (`entry_order`, `take_profit` and `stop_loss`) whereas `stop_loss` will be locked to `stop_market`. If I have to build a list with with (`entry_order` and `stop_loss`) or (`entry_order` and `trailing_stop_loss`) [of course assuming the adapter will support this scenario], I didn't find the example for this scenario.

**Comment by cjdsellers** [comment]
I see what you mean.

So as you point out, a bracket order is a specific order type which includes an attached SL and TP.

I think I understand your suggestion now that we provide an easier way of arranging contingent order(s), through the `OrderFactory`, which doesn't involve the user constructing themselves from 'raw' parts and passing the final initialized orders into an `OrderList`?

The API would probably involve passing initialized orders into some `OrderFactory` method, which would reassign the relevant fields for each order (and their `OrderInitialized`) events? e.g. something like:

```cython
order_list = order_factory.create_list(list orders)
```

```cython
order_factory.link_otos(Order parent, list oto_orders, OrderList order_list = None)
```

```cython
order_factory.link_ocos(list oco_orders, OrderList order_list = None)
```

**Comment by cjdsellers** [comment]
For your specific example, it could be:

```cython
order_list = order_factory.link_otos(entry, [stop_loss])
```

or (if you already have an `OrderList`):

```cython
order_list = order_factory.link_otos(entry, [stop_loss], order_list)
```

**Comment by rsmb7z** [comment]
> I think I understand your suggestion now that we provide an easier way of arranging contingent order(s), through the `OrderFactory`, which doesn't involve the user constructing themselves from 'raw' parts and passing the final initialized orders into an `OrderList`?

Actually opposite, fully featured way - not the easy. Easier way though, would be creating as many possibilities in the `OrderFactory` itself but I think that would be messy. To achieve like below in strategy, based on need any combination can be constructed without first creating in the `OrderFactory`. Another reason I want to construct here to make use of tags which would possibly be helpful in adapter for parsing etc.
```python
order_list_id = self.order_factory._order_list_id_generator.generate()
entry_client_order_id = self.order_factory._order_id_generator.generate()
sl_client_order_id = self.order_factory._order_id_generator.generate()

entry_order = LimitOrder(
    trader_id=self.trader_id,
    strategy_id=self.strategy_id,
    instrument_id=instrument_id,
    client_order_id=entry_client_order_id,
    order_side=order_side,
    quantity=quantity,
    price=entry_price,
    init_id=UUID4(),
    ts_init=self._clock.timestamp_ns(),
    time_in_force=time_in_force,
    expire_time_ns=0 if expire_time is None else dt_to_unix_nanos(expire_time),
    post_only=post_only_entry,
    emulation_trigger=emulation_trigger,
    contingency_type=ContingencyType.OTO,
    order_list_id=order_list_id,
    linked_order_ids=[sl_client_order_id],
    parent_order_id=None,
    tags="ENTRY",
)
sl_order = StopMarketOrder(
    trader_id=self.trader_id,
    strategy_id=self.strategy_id,
    instrument_id=entry_order.instrument_id,
    client_order_id=sl_client_order_id,
    order_side=Order.opposite_side_c(entry_order.side),
    quantity=quantity,
    trigger_price=sl_trigger_price,
    trigger_type=TriggerType.DEFAULT,
    init_id=UUID4(),
    ts_init=self._clock.timestamp_ns(),
    time_in_force=TimeInForce.GTC,
    reduce_only=True,
    emulation_trigger=emulation_trigger,
    order_list_id=order_list_id,
    parent_order_id=entry_client_order_id,
    tags="STOP_LOSS",
)
order_list = OrderList(
    order_list_id=order_list_id,
    orders=[entry_order, sl_order],
)
```

**Comment by cjdsellers** [comment]
So I've added the following:
- `OrderFactory.generate_client_order_id()` (calls internal generator)
- `OrderFactory.generate_order_list_id()` (calls internal generator)
- `OrderFactory.create_list(...)` as easier method for creating order lists

Which gets you what you wanted from the above. I'll hold off on adding any easier ways of linking orders until we learn more about the potential use cases.

Pushed to `develop`, and closing this for now.
